crates/holochain_state/src/validation_db.rs

```diff
@@ -34,3 +40,116 @@ impl ToSql for ValidationStage {
         ))
     }
 }
+
+impl FromSql for ValidationStage {
+    fn column_result(value: ValueRef<'_>) -> FromSqlResult<Self> {
+        let stage: Option<i32> = FromSql::column_result(value)?;
+        match stage {
+            None => Ok(ValidationStage::Pending),
+            Some(0) => Ok(ValidationStage::AwaitingSysDeps),
+            Some(1) => Ok(ValidationStage::SysValidated),
+            Some(2) => Ok(ValidationStage::AwaitingAppDeps),
+            Some(3) => Ok(ValidationStage::AwaitingIntegration),
+            Some(_) => Err(holochain_sqlite::rusqlite::types::FromSqlError::Other(
+                Box::new(std::io::Error::other("Invalid ValidationStage value")),
+            )),
+        }
+    }
+}
+
+/// Get the validation state of a [`DhtOp`] by its action hash and type.
+///
+/// Returns `Ok(None)` if the action is not found or there is no matching op type for the action.
+/// Otherwise, returns a combination of stage and status:
+///   - If the op has just arrived, it will return `(None, None)`.
+///   - If the op is in the process of validation, it will return `(Some(stage), None)`.
+///   - If a validation decision has been made but the op hasn't yet been integrated, it will return `(Some(stage), Some(status))`.
+///   - If the op has been fully integrated, it will return `(None, Some(status))`.
+pub async fn get_dht_op_validation_state(
+    dht_db: &holochain_sqlite::db::DbRead<DbKindDht>,
+    action_hash: ActionHash,
+    op_type: ChainOpType,
+) -> StateQueryResult<Option<(Option<ValidationStage>, Option<ValidationStatus>)>> {
+    dht_db.read_async(move |txn| -> StateQueryResult<_> {
+        let mut stmt = txn.prepare(
+            "SELECT validation_stage, validation_status, when_integrated FROM DhtOp WHERE action_hash = ? AND type = ?"
+        )?;
+
+        Ok(stmt.query_row(params![action_hash, op_type], |row| -> Result<(Option<ValidationStage>, Option<ValidationStatus>), _> {
+            let stage = row.get::<_, Option<ValidationStage>>(0)?;
+            let state = row.get::<_, Option<ValidationStatus>>(1)?;
+            Ok((stage, state))
```

Comment:
Shouldn't it check for inconsistent state like `(Some(ValidationStage::AwaitingSysDeps), Some(ValidationStatus::Valid))` ?

Reply:
The setters don't actually prevent bad states, they allow mutating these fields independently but I agree that it's not the responsibility of this code to check the values. 

My first preference would be that the database write queries only permit the allowed transitions but that isn't the case.
The next preference would be that code which mutates the state is tested and that we do have tests for in the workflows.